### ISSUE TITLE
Add Schedule-Free optimizer (online iterate averaging — anytime; Defazio et al. 2024)

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#5641]
## Summary

Odyssey ships SGD/Adam/AdamW/RMSprop/LARS/Muon-family/Lion/ADOPT/Adan/Sophia/FTRL/Shampoo/SOAP but not **Schedule-Free**. Add it as a pure-functional optimizer step under `src/odyssey/training/optimizers/`, following the first-order-moment sibling convention (`adan.mojo`, `adopt.mojo`: state in, state out; caller manages all buffers and the integer step).

Schedule-Free (Defazio et al. 2024, "The Road Less Scheduled", NeurIPS 2024, arXiv:2405.15682) replaces explicit learning-rate schedules with online iterate averaging over three coupled sequences. There is no fixed training horizon: the averaged iterate is a good checkpoint at **any** step.

## Update rule (three sequences y/z/x, per step t, 1-indexed)

```
y_t     = (1 - beta) * z_t + beta * x_t          (gradient query point)
z_{t+1} = z_t - gamma * grad(f, y_t)             (fast SGD-style sequence)
c_{t+1} = (r + 1) / (t + r + 1)                  (averaging weight)
x_{t+1} = (1 - c_{t+1}) * x_t + c_{t+1} * z_{t+1}  (running average of z)
```

- `z` (fast sequence) and `x` (running average) are the persisted buffers — same memory as SGD+momentum, no second-moment tensor.
- `y` is recomputed each step; the model trains at `y`, evaluates/checkpoints at `x` (the .train()/.eval() buffer switch).
- On step 1, initialize `z = x = params` so `y_1 = params`.
- `beta` ~ 0.9–0.95 (interpolation); `r` = weight power (r=0 → uniform average `c_{t+1} = 1/(t+1)`; larger r down-weights early iterates); `gamma` = base LR.

## Scope

- `schedule_free_step` (full hyperparameters) + `schedule_free_step_simple` (defaults beta=0.9, r=0).
- Wraps the SGD base optimizer (the `z` update is the SGD step); a different base optimizer would replace that line.
- Signature returns `(new_z, new_x, y_next)`; `params` is accepted for the shape/dtype guards and to seed step-1 state.
- float32/float64 SIMD API; raises on params/gradients shape OR dtype mismatch (no mixed dtypes).

## Acceptance criteria

- [ ] `schedule_free_step` + `schedule_free_step_simple` land under `src/odyssey/training/optimizers/`, pure-functional, mirroring adan/adopt.
- [ ] Unit test in `tests/odyssey/training/optimizers/`: shape/dtype guards; multi-step (≥3) numerical parity to 1e-9 against a committed reference so the y/z/x sequences and the evolving averaging weight c_t are exercised; simple==defaults; step-1 identity; descent direction.
- [ ] Committed Python parity reference (NumPy calculator transcribing this rule; the official `schedulefree` package is not in the pinned interpreter).
- [ ] Registered in the optimizers `__init__.mojo`, the README primitive table, and `scripts/run_primitive_test.sh`.
- [ ] `mojo format` clean, repo pre-commit clean, test compiles + passes.

## References

- Defazio, Yang, Mehta, Mishchenko, Khaled, Cutkosky (2024). The Road Less Scheduled. NeurIPS 2024. arXiv:2405.15682. https://github.com/facebookresearch/schedule_free

## Downstream

Parent of ScheduleFree+ and SF-NorMuon; combines with Prodigy. Consumed by the predictive-coding optimizer×clip×LR ablation matrix (tracking: mvillmow/Random#77).
